File: app-windows/RunnerForge.Tests/ArtifactAssertionTests.cs
using System.Formats.Tar;
using RunnerForge.Models;
using RunnerForge.Services;
using Xunit;

namespace RunnerForge.Tests;

/// <summary>
/// The artifact contract, asserted against real directories on disk.
/// </summary>
/// <remarks>
/// These tests exist because the contract's whole purpose is catching silent
/// emptiness. A gate that passes an empty artifact is worse than no gate: it
/// converts a broken build into a green run nobody looks at.
/// </remarks>
public sealed class ArtifactAssertionTests : IDisposable
{
    private readonly string _root =
        Path.Combine(Path.GetTempPath(), "rf-artifact-tests-" + Guid.NewGuid().ToString("N"));

    public ArtifactAssertionTests() => Directory.CreateDirectory(_root);

    public void Dispose()
    {
        try { if (Directory.Exists(_root)) Directory.Delete(_root, recursive: true); }
        catch (IOException) { }
    }

    private string MakeArtifact(string name, params (string RelativePath, int Bytes)[] files)
    {
        string directory = Path.Combine(_root, name);
        Directory.CreateDirectory(directory);

        foreach ((string relativePath, int bytes) in files)
        {
            string full = Path.Combine(directory, relativePath);
            Directory.CreateDirectory(Path.GetDirectoryName(full)!);
            File.WriteAllBytes(full, new byte[bytes]);
        }

        return directory;
    }

    [Fact]
    public void A_well_formed_artifact_passes()
    {
        MakeArtifact("Canary-windows-x64",
            ("VST3/Canary.vst3/Contents/x86_64-win/Canary.vst3", 1024),
            ("CLAP/Canary.clap", 2048),
            ("Standalone/Canary.exe", 4096));

        var expectation = new ArtifactExpectation
        {
            Name = "Canary-windows-x64",
            Required = true,
            Globs = ["*.vst3", "*.clap", "*.exe"],
        };

        (bool passed, long bytes, int files) = SelfTestService.VerifyArtifact(_root, expectation);

        Assert.True(passed);
        Assert.Equal(1024 + 2048 + 4096, bytes);
        Assert.Equal(3, files);
    }

    /// <summary>
    /// The exact failure `if-no-files-found: error` exists to prevent: an
    /// artifact that uploaded successfully and contains nothing.
    /// </summary>
    [Fact]
    public void An_empty_artifact_fails()
    {
        Directory.CreateDirectory(Path.Combine(_root, "Canary-windows-x64"));

        var expectation = new ArtifactExpectation
        {
            Name = "Canary-windows-x64", Required = true, Globs = ["*.vst3"],
        };

        (bool passed, _, _) = SelfTestService.VerifyArtifact(_root, expectation);
        Assert.False(passed);
    }

    [Fact]
    public void A_missing_required_artifact_fails()
    {
        var expectation = new ArtifactExpectation
        {
            Name = "Canary-macos-installer", Required = true, Globs = ["*.pkg"],
        };

        (bool passed, _, _) = SelfTestService.VerifyArtifact(_root, expectation);
        Assert.False(passed);
    }

    /// <summary>
    /// A missing OPTIONAL artifact is tolerated. This is how the AAX artifacts
    /// are legitimately absent when no SDK is configured.
    /// </summary>
    [Fact]
    public void A_missing_optional_artifact_is_tolerated()
    {
        var expectation = new ArtifactExpectation
        {
            Name = "Canary-aax-windows", Required = false, Globs = ["*.aaxplugin"],
        };

        (bool passed, _, _) = SelfTestService.VerifyArtifact(_root, expectation);
        Assert.True(passed);
    }

    [Fact]
    public void A_missing_required_glob_fails_even_when_other_files_are_present()
    {
        MakeArtifact("Canary-windows-x64",
            ("VST3/Canary.vst3", 1024),
            ("Standalone/Canary.exe", 2048));

        var expectation = new ArtifactExpectation
        {
            Name = "Canary-windows-x64", Required = true,
            Globs = ["*.vst3", "*.clap", "*.exe"],   // no .clap was produced
        };

        (bool passed, _, _) = SelfTestService.VerifyArtifact(_root, expectation);
        Assert.False(passed);
    }

    [Fact]
    public void A_present_but_zero_byte_required_file_fails()
    {
        MakeArtifact("Canary-windows-x64",
            ("VST3/Canary.vst3", 1024),
            ("CLAP/Canary.clap", 0),        // present, and useless
            ("Standalone/Canary.exe", 2048));

        var expectation = new ArtifactExpectation
        {
            Name = "Canary-windows-x64", Required = true, Globs = ["*.vst3", "*.clap", "*.exe"],
        };

        (bool passed, _, _) = SelfTestService.VerifyArtifact(_root, expectation);
        Assert.False(passed);
    }

    /// <summary>
    /// macOS bundles are tarred on the way into an artifact because
    /// upload-artifact flattens symlinks. The verifier untars before asserting.
    /// </summary>
    [Fact]
    public void A_tarred_macos_bundle_passes_after_extraction()
    {
        string staging = Path.Combine(_root, "staging");
        string bundle = Path.Combine(staging, "Canary.component");
        Directory.CreateDirectory(Path.Combine(bundle, "Contents", "MacOS"));
        File.WriteAllBytes(Path.Combine(bundle, "Contents", "MacOS", "Canary"), new byte[4096]);
        File.WriteAllText(Path.Combine(bundle, "Contents", "Info.plist"), "<plist/>");

        string artifact = Path.Combine(_root, "Canary-macos-universal");
        Directory.CreateDirectory(artifact);
        TarFile.CreateFromDirectory(staging, Path.Combine(artifact, "macos-bundles.tar"),
            includeBaseDirectory: false);

        var expectation = new ArtifactExpectation
        {
            Name = "Canary-macos-universal", Required = true, Untar = true,
            Globs = ["*.component"], Bundles = ["*.component"],
        };

        (bool passed, _, _) = SelfTestService.VerifyArtifact(_root, expectation);
        Assert.True(passed);
    }

    /// <summary>
    /// A bundle missing Contents/MacOS did not survive the round-trip and would
    /// be rejected by every host. It must not pass.
    /// </summary>
    [Fact]
    public void A_corrupted_macos_bundle_fails()
    {
        string staging = Path.Combine(_root, "staging-bad");
        string bundle = Path.Combine(staging, "Canary.component", "Contents");
        Directory.CreateDirectory(bundle);
        File.WriteAllText(Path.Combine(bundle, "Info.plist"), "<plist/>");
        // No Contents/MacOS at all.

        string artifact = Path.Combine(_root, "Canary-macos-universal");
        Directory.CreateDirectory(artifact);
        TarFile.CreateFromDirectory(staging, Path.Combine(artifact, "macos-bundles.tar"),
            includeBaseDirectory: false);

        var expectation = new ArtifactExpectation
        {
            Name = "Canary-macos-universal", Required = true, Untar = true,
            Globs = ["*.component"], Bundles = ["*.component"],
        };

        (bool passed, _, _) = SelfTestService.VerifyArtifact(_root, expectation);
        Assert.False(passed);
    }

    /// <summary>
    /// Uploading a bundle raw instead of tarring it is a specific, easy mistake.
    /// The verifier must not quietly accept it.
    /// </summary>
    [Fact]
    public void An_artifact_expecting_a_tar_but_shipped_raw_fails()
    {
        MakeArtifact("Canary-macos-universal",
            ("Canary.component/Contents/MacOS/Canary", 4096),
            ("Canary.component/Contents/Info.plist", 64));

        var expectation = new ArtifactExpectation
        {
            Name = "Canary-macos-universal", Required = true, Untar = true,
            Globs = ["*.component"], Bundles = ["*.component"],
        };

        (bool passed, _, _) = SelfTestService.VerifyArtifact(_root, expectation);
        Assert.False(passed);
    }

    // --- the contract itself -------------------------------------------------

    [Fact]
    public void The_contract_names_every_required_artifact()
    {
        IReadOnlyList<ArtifactExpectation> contract =
            ArtifactExpectation.ContractFor("Canary", aaxAvailable: false);

        string[] names = [.. contract.Select(c => c.Name)];

        Assert.Contains("Canary-windows-x64", names);
        Assert.Contains("Canary-windows-arm64", names);
        Assert.Contains("Canary-windows-installer", names);
        Assert.Contains("Canary-macos-universal", names);
        Assert.Contains("Canary-macos-installer", names);
        Assert.Contains("Canary-logs-windows", names);
        Assert.Contains("Canary-logs-macos", names);
    }

    [Fact]
    public void Without_an_sdk_the_aax_artifacts_are_optional_and_the_marker_is_required()
    {
        IReadOnlyList<ArtifactExpectation> contract =
            ArtifactExpectation.ContractFor("Canary", aaxAvailable: false);

        Assert.False(contract.First(c => c.Name == "Canary-aax-windows").Required);
        Assert.False(contract.First(c => c.Name == "Canary-aax-macos").Required);
        Assert.True(contract.First(c => c.Name == "Canary-aax-skipped").Required);
    }

    [Fact]
    public void With_an_sdk_the_aax_artifacts_are_required_and_the_marker_is_not()
    {
        IReadOnlyList<ArtifactExpectation> contract =
            ArtifactExpectation.ContractFor("Canary", aaxAvailable: true);

        Assert.True(contract.First(c => c.Name == "Canary-aax-windows").Required);
        Assert.True(contract.First(c => c.Name == "Canary-aax-macos").Required);
        Assert.False(contract.First(c => c.Name == "Canary-aax-skipped").Required);
    }

    /// <summary>
    /// Both platforms must always be represented. A run that produced only
    /// Windows output is not a successful run of this pipeline.
    /// </summary>
    [Fact]
    public void The_contract_always_requires_both_a_windows_and_a_macos_artifact()
    {
        foreach (bool aax in new[] { true, false })
        {
            IReadOnlyList<ArtifactExpectation> contract = ArtifactExpectation.ContractFor("X", aax);

            Assert.Contains(contract, c => c.Required && c.Name.Contains("-windows-"));
            Assert.Contains(contract, c => c.Required && c.Name.Contains("-macos-"));
        }
    }

    [Fact]
    public void Only_the_macos_bundle_artifact_is_untarred()
    {
        IReadOnlyList<ArtifactExpectation> contract = ArtifactExpectation.ContractFor("Canary", false);

        Assert.True(contract.First(c => c.Name == "Canary-macos-universal").Untar);
        Assert.False(contract.First(c => c.Name == "Canary-windows-x64").Untar);
        Assert.False(contract.First(c => c.Name == "Canary-macos-installer").Untar);
    }
}

File: app-macos/Tests/RunnerForgeTests/ArtifactAssertionTests.swift
import Foundation
import Testing
@testable import RunnerForge

/// The artifact contract is the deliverable. A green run that uploaded an empty
/// directory is a failure that looks like a success, so these tests build real
/// directories on disk and assert against them.
@Suite("Artifact assertion")
struct ArtifactAssertionTests {

    private func makeRoot() throws -> String {
        let root = NSTemporaryDirectory() + "rf-artifacts-\(UUID().uuidString)"
        try FileManager.default.createDirectory(atPath: root, withIntermediateDirectories: true)
        return root
    }

    private func write(_ path: String, bytes: Int) throws {
        try FileManager.default.createDirectory(
            atPath: (path as NSString).deletingLastPathComponent,
            withIntermediateDirectories: true)
        try Data(repeating: 0x41, count: bytes).write(to: URL(fileURLWithPath: path))
    }

    // -----------------------------------------------------------------------
    // The contract itself
    // -----------------------------------------------------------------------

    /// With no AAX SDK the two AAX artifacts become optional AND the skip marker
    /// becomes required, so the set is never silently one artifact short.
    @Test("the AAX rows flip together")
    func aaxRowsFlipTogether() {
        let withSdk = ArtifactExpectation.contract(for: "Canary", aaxAvailable: true)
        #expect(withSdk.first { $0.name == "Canary-aax-windows" }?.required == true)
        #expect(withSdk.first { $0.name == "Canary-aax-macos" }?.required == true)
        #expect(withSdk.first { $0.name == "Canary-aax-skipped" }?.required == false)

        let withoutSdk = ArtifactExpectation.contract(for: "Canary", aaxAvailable: false)
        #expect(withoutSdk.first { $0.name == "Canary-aax-windows" }?.required == false)
        #expect(withoutSdk.first { $0.name == "Canary-aax-macos" }?.required == false)
        #expect(withoutSdk.first { $0.name == "Canary-aax-skipped" }?.required == true)
    }

    /// macOS plugins are bundle DIRECTORIES with symlinks inside. upload-artifact
    /// flattens symlinks, so they are tarred on the way in — which means the
    /// macOS rows must say untar, or the assertion looks for a .vst3 that the
    /// artifact never contained.
    @Test("only the macOS rows untar, and they name their bundles")
    func macosRowsUntar() {
        let contract = ArtifactExpectation.contract(for: "Canary", aaxAvailable: true)
        let untarring = contract.filter(\.untar)

        #expect(untarring.map(\.name) == ["Canary-macos-universal"])
        #expect(untarring.allSatisfy { !$0.bundles.isEmpty })
        #expect(contract.filter { !$0.untar }.allSatisfy { $0.bundles.isEmpty })
    }

    @Test("every artifact name is unique and project-scoped")
    func namesAreUniqueAndScoped() {
        let contract = ArtifactExpectation.contract(for: "Canary", aaxAvailable: true)
        #expect(Set(contract.map(\.name)).count == contract.count)
        #expect(contract.allSatisfy { $0.name.hasPrefix("Canary-") })
        #expect(contract.allSatisfy { !$0.globs.isEmpty })
    }

    // -----------------------------------------------------------------------
    // Assertion against real directories
    // -----------------------------------------------------------------------

    @Test("an artifact matching its globs passes")
    func matchingArtifactPasses() async throws {
        let root = try makeRoot()
        defer { try? FileManager.default.removeItem(atPath: root) }

        let expectation = ArtifactExpectation(
            name: "Canary-windows-x64", required: true, globs: ["*.vst3", "*.clap"])

        try write("\(root)/Canary-windows-x64/Canary.vst3", bytes: 2048)
        try write("\(root)/Canary-windows-x64/Canary.clap", bytes: 1024)

        let outcome = await SelfTestService.verifyArtifact(
            downloadRoot: root, expectation: expectation,
            processRunner: ProcessRunner(logBus: LogBus()))

        #expect(outcome.passed)
        #expect(outcome.bytes == 3072)
        #expect(outcome.fileCount == 2)
    }

    /// An empty file is the failure mode that a "does it exist" check misses: a
    /// build that produced a zero-byte plugin uploads happily.
    @Test("a zero-byte artifact fails")
    func emptyArtifactFails() async throws {
        let root = try makeRoot()
        defer { try? FileManager.default.removeItem(atPath: root) }

        let expectation = ArtifactExpectation(
            name: "Canary-windows-x64", required: true, globs: ["*.vst3"])
        try write("\(root)/Canary-windows-x64/Canary.vst3", bytes: 0)

        let outcome = await SelfTestService.verifyArtifact(
            downloadRoot: root, expectation: expectation,
            processRunner: ProcessRunner(logBus: LogBus()))

        #expect(!outcome.passed)
    }

    @Test("a missing glob fails even when other files are present")
    func missingGlobFails() async throws {
        let root = try makeRoot()
        defer { try? FileManager.default.removeItem(atPath: root) }

        let expectation = ArtifactExpectation(
            name: "Canary-windows-x64", required: true, globs: ["*.vst3", "*.clap"])
        try write("\(root)/Canary-windows-x64/Canary.vst3", bytes: 2048)

        let outcome = await SelfTestService.verifyArtifact(
            downloadRoot: root, expectation: expectation,
            processRunner: ProcessRunner(logBus: LogBus()))

        #expect(!outcome.passed)
    }

    @Test("a missing required artifact fails and a missing optional one does not")
    func missingArtifact() async throws {
        let root = try makeRoot()
        defer { try? FileManager.default.removeItem(atPath: root) }

        let runner = ProcessRunner(logBus: LogBus())

        let required = await SelfTestService.verifyArtifact(
            downloadRoot: root,
            expectation: ArtifactExpectation(name: "absent", required: true, globs: ["*.vst3"]),
            processRunner: runner)
        #expect(!required.passed)

        let optional = await SelfTestService.verifyArtifact(
            downloadRoot: root,
            expectation: ArtifactExpectation(name: "absent", required: false, globs: ["*.aaxplugin"]),
            processRunner: runner)
        #expect(optional.passed)
    }

    /// The round trip the macOS rows exist for: tar the bundle, assert that
    /// Contents/MacOS and Contents/Info.plist survive.
    @Test("a tarred macOS bundle is extracted and its structure asserted")
    func tarredBundleRoundTrips() async throws {
        let root = try makeRoot()
        defer { try? FileManager.default.removeItem(atPath: root) }

        let staging = "\(root)/staging"
        let bundle = "\(staging)/Canary.vst3"
        try write("\(bundle)/Contents/MacOS/Canary", bytes: 4096)
        try write("\(bundle)/Contents/Info.plist", bytes: 256)

        let artifactDir = "\(root)/Canary-macos-universal"
        try FileManager.default.createDirectory(atPath: artifactDir, withIntermediateDirectories: true)

        let tar = Process()
        tar.executableURL = URL(fileURLWithPath: "/usr/bin/tar")
        tar.arguments = ["-cf", "\(artifactDir)/macos-bundles.tar", "-C", staging, "Canary.vst3"]
        try tar.run()
        tar.waitUntilExit()
        #expect(tar.terminationStatus == 0)

        let outcome = await SelfTestService.verifyArtifact(
            downloadRoot: root,
            expectation: ArtifactExpectation(
                name: "Canary-macos-universal", required: true,
                globs: ["*.vst3"], untar: true, bundles: ["*.vst3"]),
            processRunner: ProcessRunner(logBus: LogBus()))

        #expect(outcome.passed)
    }

    /// A tar that unpacks to a bundle with no Contents/MacOS is not a plugin,
    /// however plausible its name.
    @Test("a bundle missing Contents/MacOS fails")
    func bundleWithoutMacOsFails() async throws {
        let root = try makeRoot()
        defer { try? FileManager.default.removeItem(atPath: root) }

        let staging = "\(root)/staging"
        try write("\(staging)/Canary.vst3/Contents/Info.plist", bytes: 256)

        let artifactDir = "\(root)/Canary-macos-universal"
        try FileManager.default.createDirectory(atPath: artifactDir, withIntermediateDirectories: true)

        let tar = Process()
        tar.executableURL = URL(fileURLWithPath: "/usr/bin/tar")
        tar.arguments = ["-cf", "\(artifactDir)/macos-bundles.tar", "-C", staging, "Canary.vst3"]
        try tar.run()
        tar.waitUntilExit()

        let outcome = await SelfTestService.verifyArtifact(
            downloadRoot: root,
            expectation: ArtifactExpectation(
                name: "Canary-macos-universal", required: true,
                globs: ["*.vst3"], untar: true, bundles: ["*.vst3"]),
            processRunner: ProcessRunner(logBus: LogBus()))

        #expect(!outcome.passed)
    }

    /// A self-test that ran on GitHub's hosted runners proves nothing about this
    /// machine, so the result type says so explicitly.
    @Test("a run on hosted runners is not a valid self-test")
    func hostedRunnersDoNotCount() {
        var result = SelfTestResult(runId: 1, conclusion: "success", verdict: "pass")
        result.runnerNames = ["forge-mac-build-0", "ubuntu-24.04"]
        #expect(!result.allJobsOnForgeRunners)

        result.runnerNames = ["forge-mac-build-0", "forge-win-build-1"]
        #expect(result.allJobsOnForgeRunners)

        result.runnerNames = []
        #expect(!result.allJobsOnForgeRunners)
    }
}
